src/librustc_codegen_llvm/mir/rvalue.rs

```diff
@@ -483,8 +484,31 @@ impl<'a, 'tcx> FunctionCx<'a, 'tcx> {
                     }
                 };
                 let instance = ty::Instance::mono(bx.tcx(), def_id);
+                let fn_ty = FnType::of_instance(bx.cx, &instance);
                 let r = callee::get_fn(bx.cx, instance);
-                let val = bx.pointercast(bx.call(r, &[llsize, llalign], None), llty_ptr);
+
+                let mut llargs = Vec::with_capacity(3);
+                let indirect_ret_val = if fn_ty.ret.is_indirect() {
+                    let tmp = PlaceRef::alloca(&bx, fn_ty.ret.layout, "tmp_ret");
+                    tmp.storage_live(&bx);
+                    llargs.push(tmp.llval);
+                    Some(tmp)
+                } else {
+                    None
+                };
+                llargs.push(llsize);
+                llargs.push(llalign);
+
+                let fn_ret_val = bx.call(r, &llargs, None);
+
+                let ret_val = if let Some(ret_val) = indirect_ret_val {
+                    let o = ret_val.load(&bx).immediate();
+                    ret_val.storage_dead(&bx);
+                    o
+                } else {
+                    fn_ret_val
+                };
+                let val = bx.pointercast(ret_val, llty_ptr);
```

Comment:
What is this all doing? Why would the function being called here ever return by indirection?

Reply:
Accidentally included. Was created when I was trying to figure out how to work with the AMDGPU target machine. I tried having every function be the amdgpu kernel abi, which doesn't allow non-`void` return types.

Removed in the next version, to be pushed Soon(TM).